netlify/netlify-plugin-nextjs
# feat: added better custom header support

### Summary

Improves support for [Next.js custom headers](https://nextjs.org/docs/api-reference/next.config.js/headers). Currently the `headers()` function in next.config.js only works for [SSR](https://nextjs.org/docs/basic-features/pages#server-side-rendering)/[ISR](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration). [Headers in netlify.toml](https://docs.netlify.com/configure-builds/file-based-configuration/#headers) or a [_headers file](https://docs.netlify.com/routing/headers/#syntax-for-the-headers-file) are completely ignored for SSR/ISR. For [SSG pages](https://nextjs.org/docs/basic-features/pages#static-generation-recommended), only the headers defined in netlify.toml or a _headers file are recognized and headers defined in `headers()` in next.config.js are ignored.

Previously, you had to manually copy from the `headers()` function to netlify.toml or vice-versa so all headers were applied. Now all that is required is defining custom headers in the `headers()` function of next.config.js. The build plugin will update the Netlify configuration at build time with all the custom headers defined in the `headers()` function of next.config.js.

You can still add headers in netlify.toml or _headers, but those additional headers will only apply to SSG pages.

@KyleBlankRollins, we'll need a documentation update for this as well. Happy to write this up, just not sure how we usually work with the documentation team.

Note that ISR and SSR are not affected by this plugin fix.

### Test plan

The source code for test site can be found at https://github.com/nickytonline/next-plugin-issue-1346

Note that unless you load a page explicitly from the browser address bar, Next.js client-side routing will kick in so as you navigate to a new page, from the first-page load, you will not see the custom headers in subsequent pages that load from the client-side as the browser is not refreshing. You will need to refresh the browser to see the headers.

If you want to test the fix locally, you'll need to build the fix locally and run `npm link`, then in the test site repository, you'll need to run `npm link @netlify/plugin-nextjs`. Just a reminder that the fix can only be tested in a site that was deployed to Netlify using the plugin fix. There is no way to test this locally.

This is the configuration being used in the sites deployed in the test scenarios, https://github.com/nickytonline/next-plugin-issue-1346/blob/96471d10bb7be58287859e1e62af227eacf9e127/next.config.js

1. First check with the existing plugin that headers aren't being served. Navigate to https://62991c0c8fa1390082a46ace--the-awesome-nick-taylor-site-69bab.netlify.app/some-base-path/.
2. If you inspect the network tab for the request of the page rendered, you'll notice none of the custom headers I created are there, e.g. the `x-all-pages-header` header
3. Navigate to the deployed site that used the plugin fix, https://629984ea303dfb0ed76ba332--the-awesome-nick-taylor-site-69bab.netlify.app/some-base-path/.
4. Open the Network tab and you should see the following headers:

```
x-all-pages-header: all pages header value
x-all-pages-header2: all pages header value2
x-header-for-root-of-site: this is the header for the root of the site
x-header-for-root-of-site2: this is another the header for the root of the site
```

![Network tab showing custom headers being served from the Netlify configuration](https://user-images.githubusercontent.com/833231/171854056-66935995-425e-4a77-a0d0-bd6827792ae8.png)

5. Navigate to https://629984ea303dfb0ed76ba332--the-awesome-nick-taylor-site-69bab.netlify.app/some-base-path/en/some-path/yolo
6. Open the Network tab and you should see the following headers:

```
x-all-locales-header-some-path: all locales header some path value
x-all-pages-header: all pages header value
x-all-pages-header2: all pages header value2
x-en-locale-specific-yolo-header: en locale specific yolo header value
```

![Network tab showing custom headers being served from the Netlify configuration6](https://user-images.githubusercontent.com/833231/171854436-073a6b5f-ea72-4972-8ea8-ad37905f003b.png)

7. Navigate to https://629984ea303dfb0ed76ba332--the-awesome-nick-taylor-site-69bab.netlify.app/some-base-path/en/some-path and https://629984ea303dfb0ed76ba332--the-awesome-nick-taylor-site-69bab.netlify.app/some-base-path/es/some-path
8. Open the Network tab and you should see the following headers for each of those pages:

```
x-all-locales-header-some-path: all locales header some path value
x-all-pages-header: all pages header value
x-all-pages-header2: all pages header value2
```

![Network tab showing custom headers being served from the Netlify configuration](https://user-images.githubusercontent.com/833231/171854898-ff5ed088-4d4c-457f-912e-acc0dd0f6b25.png)

### Relevant links (GitHub issues, Notion docs, etc.) or a picture of cute animal

Closes #1346

### Standard checks:




---

🧪 Once merged, make sure to update the version if needed and that it was published correctly.

![A player heading a goal in a soccer match](https://media.giphy.com/media/lShVUmUEhwUjQ2mHym/giphy-downsized-large.gif)
